src/go/k8s/internal/client/users/client_test.go

```diff
@@ -0,0 +1,218 @@
+package users
+
+import (
+	"context"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/redpanda-data/common-go/rpadmin"
+	"github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/redpanda-data/redpanda-operator/src/go/k8s/internal/testutils"
+	"github.com/stretchr/testify/require"
+	"github.com/testcontainers/testcontainers-go/modules/redpanda"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/sasl/scram"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubectl/pkg/scheme"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestClient(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*2)
+	defer cancel()
+
+	testEnv := testutils.RedpandaTestEnv{}
+	cfg, err := testEnv.StartRedpandaTestEnv(false)
+	require.NoError(t, err)
+	require.NotNil(t, cfg)
+
+	err = v1alpha2.AddToScheme(scheme.Scheme)
+	require.NoError(t, err)
+
+	c, err := client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	require.NoError(t, err)
+	require.NotNil(t, c)
+
+	container, err := redpanda.Run(ctx, "docker.redpanda.com/redpandadata/redpanda:v23.2.8",
+		redpanda.WithEnableKafkaAuthorization(),
+		redpanda.WithEnableSASL(),
+		redpanda.WithSuperusers("user"),
+		redpanda.WithNewServiceAccount("user", "password"),
+	)
+
+	require.NoError(t, err)
+
+	broker, err := container.KafkaSeedBroker(ctx)
+	require.NoError(t, err)
+
+	admin, err := container.AdminAPIAddress(ctx)
+	require.NoError(t, err)
+
+	kafkaClient, err := kgo.NewClient(kgo.SeedBrokers(broker), kgo.SASL(scram.Auth{
+		User: "user",
+		Pass: "password",
+	}.AsSha256Mechanism()))
+	require.NoError(t, err)
+
+	rpadminClient, err := rpadmin.NewAdminAPI([]string{admin}, &rpadmin.BasicAuth{
+		Username: "user",
+		Password: "password",
+	}, nil)
+	require.NoError(t, err)
+
+	usersClient, err := NewClient(ctx, c, kadm.NewClient(kafkaClient), rpadminClient)
+	require.NoError(t, err)
+
+	for _, mechanism := range []kadm.ScramMechanism{
+		kadm.ScramSha256, kadm.ScramSha512,
+	} {
+		t.Run(mechanism.String(), func(t *testing.T) {
+			username := "testuser" + strconv.Itoa(int(time.Now().UnixNano()))
+
+			ok, err := usersClient.has(ctx, username)
+			require.NoError(t, err)
+			require.False(t, ok)
+
+			err = usersClient.create(ctx, username, "password", mechanism)
+			require.NoError(t, err)
+
+			ok, err = usersClient.has(ctx, username)
+			require.NoError(t, err)
+			require.True(t, ok)
+
+			err = usersClient.delete(ctx, username)
+			require.NoError(t, err)
+
+			ok, err = usersClient.has(ctx, username)
+			require.NoError(t, err)
+			require.False(t, ok)
+		})
+	}
+}
+
+func TestClientPasswordCreation(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*2)
```

Comment:
Nothing to change in this PR but the more restrictive timeouts here are nice. We all seemed to have pretty strong feelings about wanting `go test` to run quite fast. I generally like to have test contexts informed by the `-timeout` flag as a whole, there's a helper in the charts repo for doing exactly that.

Would it be worth standardizing on `-short` running with `-timeout ~2m` and enforcing that within CI?

Reply:
I'd be fine with that -- I've never fully run all of them locally, but I'm assuming that `go test ./...` is currently just unit tests?

That does bring up another question though, since we've been talking about making all e2e-style tests go-testable... I'm going to be toying soon with how we should go about doing some full e2e tests sans kuttl. What are your thoughts on build flagging the e2e tests or just putting them in some top-level folder that we can skip on the typical unit-based `go test` run?

Reply:
The helm charts repo is currently using `-short` to opt out of integration/e2e style tests. I'm not the biggest fan of that method, it was just the easiest to add at the time.

Build flags work pretty well but I would suggest that we use build flags to toggle the implementation of helper along the lines of: `testutil.SkipIfNotIntegration(t)`. That way e2e/integration tests will still get built, linted, and skipped by default. Flagging out an entire package gets a bit frustrating as `go test ./...` won't catch breaking changes 😅